### PR TITLE
Do not modify the outbound event.

### PR DESCRIPTION
**Pull request by n3wscott**
Do not assume json for empty content encoding. Preserve empty content encoding.

This adds but does not yet use a nop data encoder.

Fixes to tests to allow for non-defaulted app/json to be assumed.

Before doing these changes I touched the base64 encoding code. It is not quite finished.

fixes: https://github.com/cloudevents/sdk-go/issues/217
relates to https://github.com/cloudevents/sdk-go/issues/222

### CODE REVIEW

File: pkg/cloudevents/event_data.go [hunk 1]
```diff
@@ -14,6 +14,38 @@ import (
 
 // SetData implements EventWriter.SetData
 func (e *Event) SetData(obj interface{}) error {
```
Comment (alanconway) [inline pkg/cloudevents/event_data.go]: I find SetData confusing. Its behavior depends on whether SetDataContentType was called beforehand, and the case of passing []byte when contenttype is set is ambiguous: am I asking for []byte to be further encoded in the content type (e.g. "base64" for JSON) or am I providing []byte that are *already* JSON encoded? The existing API can handle both cases but it depends on setting content-type and data in the right order. 

Here's an alternative API, with a backwards-compatible SetData() implementation:

```go
package cloudevents

import (
	"context"
	"mime"

	"github.com/cloudevents/sdk-go/pkg/cloudevents/datacodec"
	"github.com/cloudevents/sdk-go/pkg/cloudevents/types"
)

// ================ A safer API: WARNING UNTESTED CODE :)

// SetDataBytes sets DataContentType=contentType, and Data=bytes.
//
// bytes must already be encoded according contentType.
//
func (e *Event) SetDataBytes(contentType string, bytes []byte) { // No error!
	e.Context.SetDataContentType(contentType)
	e.Data = bytes
}

// MarshalDataBytes sets DataContentType=contentType, marshals v as contentType,
// and sets Data=marshaled-bytes.
//
// contentType must bed a content-type supported by package datacodec.
// For other content types, encode the data elsewhere and call SetDataBytes()
//
func (e *Event) MarshalDataBytes(contentType string, v interface{}) error {
	// FIXME(alanconway) should do mime parse in datacodec.Encode(),
	// not here. The datacodec will need the full content-type to handle
	// charset etc.
	mediaType, _, err := mime.ParseMediaType(contentType)
	if err != nil {
		return err
	}
	b, err := datacodec.Encode(context.Background(), mediaType, v)
	if err != nil {
		return err
	}
	e.SetDataBytes(contentType, b)
	return err
}

// SetDataValue clears DataContentType and sets Data to one of the basic
// CloudEvents data types:
//
//    bool, int32, string, []byte, *url.URL, time.Time
//
// Data is set to types.Validate(v), but not encoded.
// When the event itself is encoded, Data will be encoded according to the
// CloudEvents default encoding rules.
//
func (e *Event) SetDataValue(v interface{}) error {
	v, err := types.Validate(v)
	if err == nil {
		e.Context.SetDataContentType("")
		e.Data = v
	}
	return err
}

// ================ Backward compatible SetData:

// SetData encodes obj as the event Data according to
// Context.GetDataContentType()
//
// If the content-type is empty, obj must be convertible to one of the
// CloudEvents basic types, and will be default encoded when the event
// is marshaled. The basic types are:
//
//    bool, int32, string, []byte, *url.URL, time.Time
//
// Changing the content-type after calling SetData may create an invalid event.
//
func (e *Event) SetData(v interface{}) error {
	if contentType := e.Context.GetDataContentType(); contentType != "" {
		return e.SetDataValue(v)
	} else {
		return e.MarshalDataBytes(contentType, v)
	}
}

```

Comment (alanconway) [reply pkg/cloudevents/event_data.go]: A big change deserves big comments :)

File: pkg/cloudevents/datacodec/nop/data.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
+package nop
```
[the remark this answers is not in the document]
Comment (n3wscott) [reply pkg/cloudevents/datacodec/nop/data.go]: I need to implement nop still.

File: pkg/cloudevents/event_data.go (hunk 1 above)
Comment (n3wscott) [reply pkg/cloudevents/event_data.go]: I agree. SetData no longer works as expected in 1.0 world.

https://github.com/cloudevents/sdk-go/issues/231

File: pkg/cloudevents/datacodec/codec.go [hunk 3]
```diff
@@ -25,13 +25,13 @@ func init() {
 	decoder = make(map[string]Decoder, 10)
 	encoder = make(map[string]Encoder, 10)
 
-	AddDecoder("", json.Decode)
+	AddDecoder("", json.Decode) // TODO: nop.Decode)
```
[the remark this answers is not in the document]
Comment (n3wscott) [reply pkg/cloudevents/datacodec/codec.go]: I will add in a followup PR

File: pkg/cloudevents/datacodec/codec_test.go [hunk 4]
```diff
@@ -113,7 +113,14 @@ func TestCodecEncode(t *testing.T) {
...
 			wantErr:     `[encode] unsupported content type: "unit/testing-invalid"`,
 		},
-
+		"blank": {
+			contentType: "",
+			in: map[string]string{
+				"a": "apple",
+				"b": "banana",
+			},
+			want: []byte(`{"a":"apple","b":"banana"}`),
```
[the remark this answers is not in the document]
Comment (n3wscott) [reply pkg/cloudevents/datacodec/codec_test.go]: This is for the data, not a CE attribute

File: pkg/cloudevents/event_data.go (hunk 1 above)
Comment (n3wscott) [reply pkg/cloudevents/event_data.go]: I think I will attempt this in a followup with adding the NOP data encoder.

File: pkg/cloudevents/datacodec/codec_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (n3wscott) [reply pkg/cloudevents/datacodec/codec_test.go]: this is still settling down, I can rework this test when we fixup the SetData

[the remark this answers is not in the document]
Comment (alanconway) [reply pkg/cloudevents/datacodec/codec_test.go]: If it's for data there has to be a content type like "text/json". If there's' no content type it has to be a CE type. Undescribed values that are not CE types are illegal.

**Comment by alanconway** [comment]
LGTM - this is moving us forward, we can raise new issues for outstanding work.

**Pull request merged**